lib/CL/devices/almaif/AlmaifDB/DBDevice.cc

```diff
@@ -0,0 +1,163 @@
+/* DBDevice.cc - Device based on parsing Almaif database and instantiating
+ *               other device types based on what it finds from there
+
+   Copyright (c) 2023 Topi Leppänen / Tampere University
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to
+   deal in the Software without restriction, including without limitation the
+   rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+   sell copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+   IN THE SOFTWARE.
+*/
+
+#include "DBDevice.hh"
+
+#include "../AlmaifShared.hh"
+
+#ifdef HAVE_XRT
+#include "../XrtDevice.hh"
+#include "../XrtRegion.hh"
+#endif
+
+DBDevice::DBDevice(const std::string &db_path) : db_(db_path) {
+
+  POCL_MSG_PRINT_INFO("STARTING INIT");
+
+  bool use_alveo_dev = false;
+  bool use_intel_dev = false;
+  if (pocl_is_option_set("XILINX_XRT")) {
+    use_alveo_dev = true;
+    used_device_type_ = AlmaifDBManager::DEVICE_TYPE::ALVEOU280;
+  }
+  if (pocl_is_option_set("INTEL_ACL")) {
+    use_intel_dev = true;
+    used_device_type_ = AlmaifDBManager::DEVICE_TYPE::ARRIA10;
+  }
+  if (use_alveo_dev && use_intel_dev) {
+    POCL_ABORT("AlmaIF: DBDevice only supports one vendor FPGA at the time\n");
+  }
+
+  std::string external_mem_params =
+      db_.externalMemoryParameters(used_device_type_);
+
+  if (use_alveo_dev) {
+    dev_ = new XrtDevice(db_.defaultKernelName(), db_.defaultBitstream(),
+                         external_mem_params, 0);
+  } else if (use_intel_dev) {
+    POCL_ABORT_UNIMPLEMENTED("AlmaIF intel device not implemented\n");
+  } else {
+    POCL_ABORT("AlmaIF: DBDevice didn't find any vendor FPGAs\n");
+  }
+
+  ControlMemory = dev_->ControlMemory;
+  InstructionMemory = dev_->InstructionMemory;
+  CQMemory = dev_->CQMemory;
+  DataMemory = dev_->DataMemory;
+  RelativeAddressing = dev_->RelativeAddressing;
+  HasHardwareClock = dev_->HasHardwareClock;
+  HwClockFrequency = dev_->HwClockFrequency;
+  PointerSize = dev_->PointerSize;
+  ExternalMemory = dev_->ExternalMemory;
+  AllocRegions = dev_->AllocRegions;
+}
+
+DBDevice::~DBDevice() { delete dev_; }
+
+void DBDevice::programBIKernelBitstream(BuiltinKernelId bik_id) {
```

Comment:
There are no comments at all in this file. Please add at least method comments to non-self-clear methods.

Reply:
added